Redesign group settings UI

m-e-l-u-h-a-n opened the pull request:
<!-- Describe your pull request here.-->

Fixes: Adds preparatory commits to address first two points from https://github.com/zulip/zulip/issues/19526.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![group-ui](https://user-images.githubusercontent.com/63504956/185799380-21f6a39a-337a-4a73-88a0-2d016e9359b3.gif)


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

m-e-l-u-h-a-n commented:
Tagging @salmonax for peer reviews 🙂️.

timabbott commented:
Thinking more about variable names; do we want `static/js/user_group_settings.js` as the name for the new overlay's main module? I'm thinking about the parallel with `static/js/stream_settings.js`, and how it's been nice to have the shared `static/templates/stream_settings/` directory for templates that make up that screen.

m-e-l-u-h-a-n commented:
> Thinking more about variable names; do we want `static/js/user_group_settings.js` as the name for the new overlay's main module? I'm thinking about the parallel with `static/js/stream_settings.js`, and how it's been nice to have the shared `static/templates/stream_settings/` directory for templates that make up that screen.

I've opened [#design > User groups UI](https://chat.zulip.org/#narrow/stream/101-design/topic/User.20groups.20UI) for discussing this.

sahil839 commented on static/js/user_groups_settings_ui.js:
```diff
@@ -1,14 +1,35 @@
 
 export function setup_page(callback) {
     function populate_and_fill() {
+        $("#manage_groups_container").empty();
```
Shouldn't this be added in previous commit itself?

timabbott replied:
Yeah, and usually it's a better pattern to just do `.html()` rather than `.empty` + `.append`.

sahil839 reviewed (commented):
Thanks for working on this. Overall looks good as an initial prototype.

I tested it and noticed a couple of things.
- The subscribers list is not cleared after I create a user group.
- There is no border at the bottom of the members table.
- Do we want to refer participants of groups as "Subscribers"/"Members"? Currently both are being used.

Also, would be good if you share screenshots/GIF as it would make it easy to review the UI.

timabbott commented:
I did a review of this and generally it looks like a great checkpoint. I pushed back a version that adds a couple new block comments in the last commit in the file that we've duplicated and will be low-hanging fruit to deduplicate later.

> Do we want to refer participants of groups as "Subscribers"/"Members"? Currently both are being used.

I think "Members" is the term we want to use for people in a user group; we've formalized that pretty strongly in both the UI and API. So I think that suggests a few changes:
* `user_group_create_subscribers.js` => `user_group_create_members.js`
* `user_group_create_subscribers_data.js` => `user_group_create_members_data.js`

While we're at it, what do you think about renaming
* `user_group_settings_ui.js` => `user_groups_overlay.js`
* `user_group_settings/user_groups_settings_overlay.hbs` => `user_groups/user_groups_overlay.hbs`
* `static/templates/` => `static/templates/user_groups/`

I feel like that feels more readable; and then we'd name a module for picking user groups (for use in other settings UI, say) something like `user_groups_picker.js`. We'd likely eventually want to rename the streams and settings modules to have a similar naming scheme.

Aside from these naming things, I'd be happy to merge this and plan to build forward from here, since as I mentioned, it's a good checkpoint, and I think most tweaks we'll want to make will be nicer. The one thing that we should do to support that is adding an "if page_params.development_environment" check on the `hashchange.js` logic, i.e. have this component be inaccessible in production until we're ready to get more eyes on it on chat.zulip.org.

@m-e-l-u-h-a-n does that sound good as a plan for this PR? If so, please send me a PM in chat.zulip.org once you've checked off these things and I can merge this checkpoint (it'll really help me avoid rereading the 700 lines of this that we already have when it comes time to review the next block of functionality).

m-e-l-u-h-a-n commented:
Thanks for the reviews, I've updated the PR with following changes.

- Added a commit to have a very basic setup of right panel for group edit (perliminary enough to be able to write  proper logic for hashchange.)
- Added commit to address different cases of hashchange for group settings overlay. Also added a check for `page_params.development_environment' to hide the overlay in deployment.
- Clear member list once the group once the group is created.
- Renamed classes and files wherever possible to formalize members instead of streams for user group.

**Following part is still left from previous reviews:**
- Following renames:
   - `user_group_settings_ui.js` => `user_groups_overlay.js`
   - `user_group_settings/user_groups_settings_overlay.hbs` => `user_groups/user_groups_overlay.hbs`
   - `static/templates/` => `static/templates/user_groups/`

  These are left because we initially decided to keep names analogus to stream settings files, it will also be helpful if later on decide to have a common UI setup for both it would be easy to search and refactor similar files. So I would wait for confirmation over these renames.
- `There is no border at the bottom of the members table`, I am still trying to figure out appropriate changes for this and will complete this while I wait for confirmation on above point.

m-e-l-u-h-a-n commented:
About using member instead of subscriber, we have term subscriber used a few times (8-10) in [static/js/user_group_create_members.js](https://github.com/zulip/zulip/pull/22214/commits/4c30be16acfab8fd569b1015048b424ba06c01fd#diff-854c2db9b7883124c6da58878104230d9adb73a01ed93d797e689a7998d54609) because I reused following templates because they were exact copies:
- `/stream_settings/new_stream_user.hbs` for user row in subscriber table.
- `/stream_settings/add_subscribers_form` to add pill widget for adding members in the table

Should I create a separate files for user groups similar to above `.hbs` modules to have naming consistency of `members` vs `subscriber`. Similar repetition of using subscriber instead of members will also occur in user groups edit settings if we do not have separate files.

timabbott commented:
I think it's totally fine to have a few uses of "subscriber" in the implementation where that makes it easier to share code or verify code is analogous between related components.

m-e-l-u-h-a-n commented:
@timabbott  any suggestions about left file renames, mentioned in https://github.com/zulip/zulip/pull/22214#issuecomment-1221564213.

timabbott commented:
I've merged this, after adding a comment to explain the `development_environment` check in the `hashchange` logic.

Regarding the `subscribers` / `members` naming, I think what you have here is pretty reasonable. I suspect what we will end up wanting to do to limit the number of places we have "subscribers" appear in the "user_groups" codebase would be to have them both use a common component with a name like `user_collection_picker`  (haven't thought about what's best) that expresses the idea of what we're doing, and then it's very reasonable that we'd use such a picker to select either subscribers or members.

timabbott commented on static/styles/subscriptions.css:
```diff
@@ -466,13 +468,14 @@ h4.stream_setting_subsection_title {
         and its styling similar to streams list.
     */
-    height: 100% !important;
+    height: calc(100% - 42px) !important;
```
So I probably should have brought this up before merging, but didn't notice it in time. 

This appears to be a change to the "stream name" search as well. Is it fixing a bug in that context, or introducing one?

m-e-l-u-h-a-n replied:
This rule would have no effect on stream settings overlay. It is applied here because stream settings have an extra row of widgets for (`subscribed | all`  `+`) . So the rule for height defined on [subscription.css #L469-L476](https://github.com/zulip/zulip/blob/main/static/styles/subscriptions.css#L469-L476) works well for stream list on stream settings overlay. I just tweaked the same rule for user groups as we have just one widget (search) row. I have marked it as temporary rule in the comments because it would be removed and we would be able to use height rule defined in [subscription.css #L469-L476](https://github.com/zulip/zulip/blob/main/static/styles/subscriptions.css#L469-L476) once  we have similar widget row for user group settings as well.

timabbott commented on static/templates/user_group_settings/user_group_creation_form.hbs:
```diff
@@ -0,0 +1,35 @@
+            </section>
+            <section class="block">
+                <label for="create_user_group_description">
+                    {{t "User group description" }}
```
I feel like "User group description" feels overly wordy. @alya thoughts on changing this label and the similar "Stream description" label in "Manage streams" to just "Description"? 

I think there's a case for having the form still use "User group name" / "Stream name" at the very top, since it helps remind you what you're doing -- just having "Name" there could confuse someone new to the product about what they're naming.

alya replied:
Sure, I'm happy to have it changed to "Description" as long as we do this for both groups and streams. Looks like it won't require any documentation updates.
